File: scripts/build-safmr.mjs
// Build a static voucher-rent dataset (ZIP → [0BR..4BR]) from the HUD FMR API.
//
// Two passes, because HUD only publishes ZIP-level Small Area FMR for ~38 metros:
//   1. SAFMR pass — iterate every HUD metro, collect per-ZIP SAFMR rows. This
//      sidesteps the unreliable ZIP→metro crosswalk by inverting metro→ZIPs.
//   2. County-FMR pass — for the curated markets (assumptions.ts) whose ZIPs
//      aren't SAFMR, look up the market's COUNTY FMR via /fmr/statedata (keyed
//      by the market's declared county — no crosswalk) and assign it to those
//      ZIPs. SAFMR (pass 1) always wins where both exist.
//
//   HUDUSER_API_KEY=… node scripts/build-safmr.mjs [year]
//
// Writes src/lib/hud/safmr-data.json → { fiscalYear, zips: { zip: [..5..] } }
import { writeFileSync, mkdirSync, readFileSync } from "node:fs";
import { dirname, resolve } from "node:path";
import { fileURLToPath } from "node:url";

const KEY = process.env.HUDUSER_API_KEY;
if (!KEY) { console.error("HUDUSER_API_KEY missing"); process.exit(1); }
const YEAR = Number(process.argv[2]) || 2026;
const FALLBACK_YEAR = YEAR - 1;
const headers = { Authorization: `Bearer ${KEY}` };
const BASE = "https://www.huduser.gov/hudapi/public/fmr";
const __dirname = dirname(fileURLToPath(import.meta.url));
const sleep = (ms) => new Promise((r) => setTimeout(r, ms));
const BR_COLS = ["Efficiency", "One-Bedroom", "Two-Bedroom", "Three-Bedroom", "Four-Bedroom"];

async function getJson(url, tries = 7) {
  for (let i = 0; i < tries; i++) {
    try {
      const res = await fetch(url, { headers });
      if (res.status === 429 || res.status >= 500) { await sleep(Math.min(9000, 700 * 2 ** i)); continue; }
      if (!res.ok) return { __status: res.status };
      return await res.json();
    } catch { await sleep(Math.min(9000, 700 * 2 ** i)); }
  }
  return { __status: 0 };
}

const rents = (r) => BR_COLS.map((c) => Number(r[c]) || 0);
const valid = (a) => a.some((v) => v > 0);

const zips = {};

// ── Pass 1: SAFMR per-ZIP ────────────────────────────────────────────────────
const meta = await getJson(`${BASE}/listMetroAreas`);
const metros = (Array.isArray(meta) ? meta : meta.data || []).map((m) => m.cbsa_code).filter(Boolean);
console.log(`Pass 1 — SAFMR per-ZIP across ${metros.length} metros (FY${YEAR})…`);
const BATCH = 8;
for (let i = 0; i < metros.length; i += BATCH) {
  await Promise.all(metros.slice(i, i + BATCH).map(async (code) => {
    let data = await getJson(`${BASE}/data/${code}?year=${YEAR}`);
    if (data.__status) data = await getJson(`${BASE}/data/${code}?year=${FALLBACK_YEAR}`);
    const bd = data?.data?.basicdata;
    for (const r of Array.isArray(bd) ? bd : bd ? [bd] : []) {
      const zip = String(r.zip_code || "");
      if (/^\d{5}$/.test(zip) && !zips[zip] && valid(rents(r))) zips[zip] = rents(r);
    }
  }));
  await sleep(100);
}
const safmrCount = Object.keys(zips).length;
console.log(`  SAFMR ZIPs: ${safmrCount}`);

// ── Pass 2: county FMR for curated markets ───────────────────────────────────
const src = readFileSync(resolve(__dirname, "../src/lib/config/assumptions.ts"), "utf8");
const re = /id:\s*"([^"]+)",\s*label:\s*"([^"]+)",\s*state:\s*"([^"]+)",\s*county:\s*"([^"]+)"[\s\S]*?zips:\s*\[([^\]]*)\]/g;
const markets = [];
let m;
while ((m = re.exec(src))) {
  markets.push({ id: m[1], label: m[2], state: m[3], county: m[4], zips: (m[5].match(/\d{5}/g) || []) });
}
await sleep(5000); // cool-down so statedata isn't rate-limited after pass 1's 639 calls
console.log(`Pass 2 — county FMR for ${markets.length} curated markets…`);

const stateCache = {};
async function stateData(st) {
  if (!stateCache[st]) {
    let d = await getJson(`${BASE}/statedata/${st}?year=${YEAR}`);
    if (!d?.data?.counties) d = await getJson(`${BASE}/statedata/${st}?year=${FALLBACK_YEAR}`);
    stateCache[st] = { counties: d?.data?.counties || [], metros: d?.data?.metroareas || [] };
  }
  return stateCache[st];
}
const norm = (s) =>
  String(s || "").toLowerCase().replace(/\s+(county|parish|borough|city|msa|hud metro fmr area)\b.*$/i, "").trim();

const unmatched = [];
for (const mk of markets) {
  const missing = mk.zips.filter((z) => !zips[z]);
  if (!missing.length) continue;
  const { counties, metros } = await stateData(mk.state);
  const want = norm(mk.county);
  const city = norm(mk.label.replace(/\s*\(.*$/, "")); // "Macon (Bibb)" -> "macon"
  const row =
    counties.find((c) => norm(c.county_name) === want) ||
    counties.find((c) => { const n = norm(c.county_name); return n && (n.includes(want) || want.includes(n)); }) ||
    metros.find((m) => { const n = norm(m.metro_name); return n && (n.includes(city) || n.includes(want)); });
  if (!row) { unmatched.push(`${mk.label} (${mk.state}/${mk.county})`); continue; }
  const fmr = rents(row);
  if (!valid(fmr)) { unmatched.push(`${mk.label} (zero FMR)`); continue; }
  for (const z of missing) zips[z] = fmr;
}

console.log(`  county-FMR ZIPs added: ${Object.keys(zips).length - safmrCount}`);
if (unmatched.length) console.log(`  ⚠ unmatched markets: ${unmatched.join("; ")}`);

const out = resolve(__dirname, "../src/lib/hud/safmr-data.json");
mkdirSync(dirname(out), { recursive: true });
writeFileSync(out, JSON.stringify({ fiscalYear: YEAR, zips }));
console.log(`\nDone. ${Object.keys(zips).length} ZIPs total → ${out}`);

File: scripts/build-safmr.py
#!/usr/bin/env python3
"""Build the static voucher-rent dataset from HUD's OFFICIAL FY2026 SAFMR file.

The HUD live API is too unstable for this (its /fmr/data endpoint intermittently
404s its own FY2026 data, ZIP->metro crosswalks mismatch CBSA vintages). HUD's
published SAFMR spreadsheet, by contrast, is the authoritative, complete,
deterministic source — every SAFMR ZIP with FY2026 0-4BR rates.

  HUDUSER_API_KEY=...  python3 scripts/build-safmr.py [year]

Writes src/lib/hud/safmr-data.json  ->  { fiscalYear, source, zips: { zip: [..5..] } }
Pass 1: parse HUD's SAFMR-by-ZIP file (no API).
Pass 2: county FMR via /fmr/statedata for any curated ZIP not in the SAFMR file
        (rare — SAFMR areas cover almost all curated markets).
"""
import io, json, os, re, sys, urllib.request
import openpyxl

YEAR = int(sys.argv[1]) if len(sys.argv) > 1 else 2026
ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
OUT = os.path.join(ROOT, "src/lib/hud/safmr-data.json")
UA = "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0 Safari/537.36"
SAFMR_URL = f"https://www.huduser.gov/portal/datasets/fmr/fmr{YEAR}/fy{YEAR}_safmrs_revised.xlsx"
API = "https://www.huduser.gov/hudapi/public/fmr"


def fetch(url, headers):
    return urllib.request.urlopen(urllib.request.Request(url, headers=headers), timeout=120).read()


# ── Pass 1: official SAFMR spreadsheet ───────────────────────────────────────
print(f"Downloading HUD FY{YEAR} SAFMR file…")
xlsx = fetch(SAFMR_URL, {"User-Agent": UA, "Accept": "*/*",
                         "Referer": "https://www.huduser.gov/portal/datasets/fmr/smallarea/index.html"})
ws = openpyxl.load_workbook(io.BytesIO(xlsx), read_only=True)["SAFMRs"]
it = ws.iter_rows(values_only=True)
next(it)  # header
zips = {}
# Base SAFMR (40th-pct) columns: 0BR=3, 1BR=6, 2BR=9, 3BR=12, 4BR=15 (skip the
# 90%/110% payment-standard columns between them).
COLS = [3, 6, 9, 12, 15]
for r in it:
    z = str(r[0]).strip() if r[0] is not None else ""
    if z.isdigit() and len(z) < 5:
        z = z.zfill(5)
    if not re.fullmatch(r"\d{5}", z):
        continue
    try:
        vals = [int(round(float(r[c]))) for c in COLS]
    except (TypeError, ValueError):
        continue
    if all(v > 0 for v in vals) and z not in zips:
        zips[z] = vals
print(f"  SAFMR ZIPs from file: {len(zips)}")

# ── Pass 2: county FMR fallback for curated ZIPs not in the SAFMR file ────────
src = open(os.path.join(ROOT, "src/lib/config/assumptions.ts")).read()
markets = re.findall(
    r'state:\s*"([^"]+)",\s*county:\s*"([^"]+)"[\s\S]*?zips:\s*\[([^\]]*)\]', src)
key = os.environ.get("HUDUSER_API_KEY")
state_cache = {}


def statedata(st):
    if st not in state_cache:
        d = json.loads(fetch(f"{API}/statedata/{st}?year={YEAR}", {"Authorization": f"Bearer {key}"}))
        state_cache[st] = (d.get("data", {}).get("counties", []), d.get("data", {}).get("metroareas", []))
    return state_cache[st]


def norm(s):
    return re.sub(r"\s+(county|parish|borough|city|msa|hud metro fmr area)\b.*$", "", (s or "").lower()).strip()


added, unmatched = 0, []
for state, county, zipstr in markets:
    zs = re.findall(r"\d{5}", zipstr)
    missing = [z for z in zs if z not in zips]
    if not missing or not key:
        continue
    counties, metros = statedata(state)
    want = norm(county)
    row = next((c for c in counties if norm(c["county_name"]) == want), None) \
        or next((c for c in counties if want in norm(c["county_name"]) or norm(c["county_name"]) in want), None) \
        or next((m for m in metros if want in norm(m["metro_name"])), None)
    if not row:
        unmatched.append(f"{state}/{county}")
        continue
    fmr = [int(row[k]) for k in ["Efficiency", "One-Bedroom", "Two-Bedroom", "Three-Bedroom", "Four-Bedroom"]]
    if all(v > 0 for v in fmr):
        for z in missing:
            zips[z] = fmr
            added += 1
print(f"  county-FMR ZIPs added: {added}" + (f"  ⚠ unmatched: {unmatched}" if unmatched else ""))

os.makedirs(os.path.dirname(OUT), exist_ok=True)
json.dump({"fiscalYear": YEAR, "source": f"HUD FY{YEAR} SAFMRs (revised) + county FMR fallback", "zips": zips},
          open(OUT, "w"))
print(f"\nDone. {len(zips)} ZIPs → {OUT}")
